File: fast-tax-python/REST/get_best_match_rest.py
from ft_response import GetBestMatchResponse, BestMatchTaxInfo, InformationComponent, Error
import requests

# Endpoint URLs for ServiceObjects FastTax (FT) API
primary_url = "https://sws.serviceobjects.com/ft/web.svc/json/GetBestMatch?"
backup_url = "https://swsbackup.serviceobjects.com/ft/web.svc/json/GetBestMatch?"
trial_url = "https://trial.serviceobjects.com/ft/web.svc/json/GetBestMatch?"

def get_best_match(
    address: str,
    address2: str,
    city: str,
    state: str,
    zip: str,
    tax_type: str,
    license_key: str,
    is_live: bool = True
) -> GetBestMatchResponse:
    """
    Call ServiceObjects FastTax (FT) API's GetBestMatch endpoint
    to retrieve tax rate information (e.g., total tax rate, city, county, state rates) for a given US address.

    Parameters:
        address: Address line of the address to get tax rates for (e.g., "123 Main Street").
        address2: Secondary address line (e.g., "Apt 4B"). Optional.
        city: The city of the address (e.g., "New York"). Optional if zip is provided.
        state: The state of the address (e.g., "NY"). Optional if zip is provided.
        zip: The ZIP code of the address. Optional if city and state are provided.
        tax_type: The type of tax to look for ("sales" or "use").
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.

    Returns:
        GetBestMatchResponse: Parsed JSON response with tax rate results or error details.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """
    params = {
        "Address": address,
        "Address2": address2,
        "City": city,
        "State": state,
        "Zip": zip,
        "TaxType": tax_type,
        "LicenseKey": license_key,
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error')
        if not (error is None or error.get('Number') != "4"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"FastTax service error: {data['Error']}")
            else:
                # Trial mode error is terminal
                raise RuntimeError(f"FastTax trial error: {data['Error']}")

        # Convert JSON response to GetBestMatchResponse for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None

        return GetBestMatchResponse(
            TaxInfoItems=[
                BestMatchTaxInfo(
                    Zip=ti.get("Zip"),
                    City=ti.get("City"),
                    County=ti.get("County"),
                    StateAbbreviation=ti.get("StateAbbreviation"),
                    StateName=ti.get("StateName"),
                    TaxRate=ti.get("TaxRate"),
                    StateRate=ti.get("StateRate"),
                    CityRate=ti.get("CityRate"),
                    CountyRate=ti.get("CountyRate"),
                    CountyDistrictRate=ti.get("CountyDistrictRate"),
                    CityDistrictRate=ti.get("CityDistrictRate"),
                    SpecialDistrictRate=ti.get("SpecialDistrictRate"),
                    InformationComponents=[
                        InformationComponent(Name=comp.get("Name"), Value=comp.get("Value"))
                        for comp in ti.get("InformationComponents", [])
                    ] if "InformationComponents" in ti else [],
                    TotalTaxExempt=ti.get("TotalTaxExempt"),
                    NotesCodes=ti.get("NotesCodes"),
                    NotesDesc=ti.get("NotesDesc")
                )
                for ti in data.get("TaxInfoItems", [])
            ] if "TaxInfoItems" in data else [],
            MatchLevel=data.get("MatchLevel"),
            Error=error,
            Debug=data.get("Debug", [])
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"FastTax backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None

                return GetBestMatchResponse(
                    TaxInfoItems=[
                        BestMatchTaxInfo(
                            Zip=ti.get("Zip"),
                            City=ti.get("City"),
                            County=ti.get("County"),
                            StateAbbreviation=ti.get("StateAbbreviation"),
                            StateName=ti.get("StateName"),
                            TaxRate=ti.get("TaxRate"),
                            StateRate=ti.get("StateRate"),
                            CityRate=ti.get("CityRate"),
                            CountyRate=ti.get("CountyRate"),
                            CountyDistrictRate=ti.get("CountyDistrictRate"),
                            CityDistrictRate=ti.get("CityDistrictRate"),
                            SpecialDistrictRate=ti.get("SpecialDistrictRate"),
                            InformationComponents=[
                                InformationComponent(Name=comp.get("Name"), Value=comp.get("Value"))
                                for comp in ti.get("InformationComponents", [])
                            ] if "InformationComponents" in ti else [],
                            TotalTaxExempt=ti.get("TotalTaxExempt"),
                            NotesCodes=ti.get("NotesCodes"),
                            NotesDesc=ti.get("NotesDesc")
                        )
                        for ti in data.get("TaxInfoItems", [])
                    ] if "TaxInfoItems" in data else [],
                    MatchLevel=data.get("MatchLevel"),
                    Error=error,
                    Debug=data.get("Debug", [])
                )
            except Exception as backup_exc:
                raise RuntimeError("FastTax service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"FastTax trial error: {str(req_exc)}") from req_exc

File: fast-tax-nodejs/REST/get_best_match_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { GetBestMatchResponse } from './ft_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects FastTax (FT) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/ft/web.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects FastTax (FT) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/ft/web.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects FastTax (FT) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/ft/web.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.Number is not equal to '4'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.Number !== '4';

/**
 * <summary>
 * Constructs a full URL for the GetBestMatch API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}GetBestMatch?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<GetBestMatchResponse>">A promise that resolves to a GetBestMatchResponse object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new GetBestMatchResponse(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects FastTax (FT) API's GetBestMatch endpoint,
 * retrieving tax rate information (e.g., total tax rate, city, county, state rates) for a given US address
 * with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const GetBestMatchClient = {
    /**
     * <summary>
     * Asynchronously invokes the GetBestMatch API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.Number == '4') in live mode.
     * </summary>
     * @param {string} Address - Address line of the address to get tax rates for (e.g., "123 Main Street").
     * @param {string} Address2 - Secondary address line (e.g., "Apt 4B"). Optional.
     * @param {string} City - The city of the address (e.g., "New York"). Optional if zip is provided.
     * @param {string} State - The state of the address (e.g., "NY"). Optional if zip is provided.
     * @param {string} Zip - The ZIP code of the address. Optional if city and state are provided.
     * @param {string} TaxType - The type of tax to look for ("sales" or "use").
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<GetBestMatchResponse>} - A promise that resolves to a GetBestMatchResponse object.
     */
    async invokeAsync(Address, Address2, City, State, Zip, TaxType, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Address,
            Address2,
            City,
            State,
            Zip,
            TaxType,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the GetBestMatch API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} Address - Address line of the address to get tax rates for (e.g., "123 Main Street").
     * @param {string} Address2 - Secondary address line (e.g., "Apt 4B"). Optional.
     * @param {string} City - The city of the address (e.g., "New York"). Optional if zip is provided.
     * @param {string} State - The state of the address (e.g., "NY"). Optional if zip is provided.
     * @param {string} Zip - The ZIP code of the address. Optional if city and state are provided.
     * @param {string} TaxType - The type of tax to look for ("sales" or "use").
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {GetBestMatchResponse} - A GetBestMatchResponse object with tax rate details or an error.
     */
    invoke(Address, Address2, City, State, Zip, TaxType, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Address, Address2, City, State, Zip, TaxType, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { GetBestMatchClient, GetBestMatchResponse };